aclk/agent_cloud_link.c

```diff
@@ -133,6 +141,41 @@ struct aclk_query_queue {
     u_int64_t count;
 } aclk_queue = { .aclk_query_head = NULL, .aclk_query_tail = NULL, .count = 0 };
 
+/*
+ * After a connection failure -- delay in milliseconds
+ * When a connection is established, the delay function
+ * should be called with
+ *
+ * mode 0 to reset the delay
+ * mode 1 to sleep for the calculated amount of time [0 .. ACLK_MAX_BACKOFF_DELAY * 1000] ms
+ *
+ */
+unsigned long int aclk_delay(int mode)
+{
+    static  int fail = -1;
+    unsigned long int delay;
+
+    if (!mode || fail == -1) {
+        srandom(time(NULL));
+        fail = mode-1;
+        return 0;
+    }
+
+    delay = (1 << fail);
+
+    if (delay >= ACLK_MAX_BACKOFF_DELAY) {
+        delay = ACLK_MAX_BACKOFF_DELAY * 1000;
+    }
+    else {
+        fail++;
+        delay = (delay * 1000) + (random() % 1000);
```

Comment:
Why are you adding jitter to the back-off times? I've seen this in services that being started at the same time (e.g. Lets Encrypt renewals being triggered in crontabs) but in this case I don't understand what the common start time would be.

Reply:
If the link to our cloud infrastructure drops when we have 10000 agents connected, they will all retry in 0, 1, 2, 4 .. etc seconds. So we will get 10000 attempts at the same time. The jitter will smooth that out. I put a link in a previous comment -- recommendation for IoT devices. 

Whether it will help or not -- I can't say :smile:

Reply:
That is a really good point - but is one second of jitter enough to help in that scenario? This is something that should have been groomed with @cosmix and the cloud team.

Reply:
@stelfrag followed these guidelines https://cloud.google.com/iot/docs/how-tos/exponential-backoff
```
random_number_milliseconds is a random number of milliseconds less than or equal to 1000
```

Reply:
Good point. If there is capacity planning already in place i.e. that we know we will handle X number of requests / second and we have a good estimation of agents claimed. The latter I think is unknown. It can be discussed of course and adjust the value as needed (either statically,  via agent configuration or cloud-to-agent setting)